### PR TITLE
Distinguish missing from corrupted data

### LINKED ISSUE #64: Decoder needs to distinguish between "needs more data" and "corrupted stream"
Currently it just returns a `()` error type, so we cannot tell apart these two cases. This means that there's no way to detect stream errors.

The decoder can use a library like anyhow to provide rich text-based errors too.

### PR DESCRIPTION
**Pull request by jonas-schievink**
Replaces the `()` error type with a simple enum that distinguishes missing data from corrupted data in the decoder.

Fixes #64

### CODE REVIEW

**Review by japaric — COMMENTED** [review]
The `From` implementation to go from `io::Error` to `DecodeError` can map to any of its 2 variants but there are instances in this code where `io::Error` is always mapped to `UnexpectedEof`. Should we use `into` in those cases?

File: decoder/src/lib.rs [hunk 1]
```diff
@@ -280,8 +320,11 @@ const MAX_NUM_BOOL_FLAGS: usize = 8;
 
 impl<'t, 'b> Decoder<'t, 'b> {
     /// Reads a byte of packed bools and unpacks them into `args` at the given indices.
-    fn read_and_unpack_bools(&mut self) -> Result<(), ()> {
-        let bool_flags = self.bytes.read_u8().map_err(drop)?;
+    fn read_and_unpack_bools(&mut self) -> Result<(), DecodeError> {
+        let bool_flags = self
+            .bytes
+            .read_u8()
```
[the remark this answers is not in the document]
Comment (jonas-schievink) [reply decoder/src/lib.rs:326]: Oof, yeah, I added the `From` impl after I was already halfway through

File: decoder/src/lib.rs [hunk 2]
```diff
@@ -231,6 +232,43 @@ struct FormatSliceElement<'t> {
...
+    }
+}
+
+impl Error for DecodeError {}
+
+fn read_leb128(bytes: &mut &[u8]) -> Result<u64, DecodeError> {
+    match leb128::read::unsigned(bytes) {
+        Ok(val) => Ok(val),
+        Err(leb128::read::Error::Overflow) => Err(DecodeError::Malformed),
+        Err(leb128::read::Error::IoError(_)) => Err(DecodeError::UnexpectedEof),
```
[the remark this answers is not in the document]
Comment (jonas-schievink) [reply decoder/src/lib.rs]: Yes, fixed

**Review by japaric — APPROVED** [review]
Thanks! Looks good to me

**Pull request merged**